wordpress-mobile/WordPress-Android
# Feature: IN-APP UPDATES

This PR implements the [In-app updates feature](https://jetpackmobile.wordpress.com/2024/05/08/pt-in-app-updates/) on Android.

More details about Google's in-app updates can be found [here](https://developer.android.com/guide/playcore/in-app-updates). Please read the contents of the link first to get more context. We use the Google Play Core libraries to support UX flows for in-app updates, allowing users to keep our app up to date on their devices.

## To Test:

Testing this feature requires many steps. We need to test both the "immediate update" flow and the "flexible update" flow.

When an update is available, we need to decide which flow to use. We utilize a remote config value from the backend representing the last version of our app that requires an immediate update. This is usually a version that contains a critical bug fix or a new feature that our app cannot function properly without. If the current version of the installed app on the user's device is lower than the remote config value, then the immediate update flow is launched. Otherwise, we go with the flexible one.

**ATTENTION:** The current remote config value is 80000. This means that when you try to update a version of the app that is < 80000, the immediate flow will be launched.

To test this feature, we use Internal app sharing as Google suggests [here](https://developer.android.com/guide/playcore/in-app-updates/test).

Please read how Internal app sharing works [here](https://support.google.com/googleplay/android-developer/answer/9844679).

**ATTENTION:** As app bundle links in Internal app sharing might expire, you may need to reupload an app bundle to get a new link to use. The page for uploading is [here](https://play.google.com/console/internal-app-sharing). If you have the [Release apps to testing tracks](https://support.google.com/googleplay/android-developer/answer/9844686#release_testing) permission, you’re authorized to upload app bundles and APKs for internal sharing by default. If you don't, you can make a request in Systems P2 like [this](https://wp.me/pMz3w-jQU).

1. Enable internal app sharing on your device following the instructions:
```
Open the Google Play Store app Google Play.
Tap Menu Menu > Settings.
In the "About" section, tap the Play Store version 7 times.
After the Internal app sharing setting appears, tap the switch to turn on internal app sharing.
Tap Turn on.  
```
2. Download the 3 app versions on your computer. Having these versions on your machine will allow you to reupload them as many times as you want and get new links if the old ones have expired. Link [here](https://drive.google.com/drive/folders/1GuAC0m707sLxzIdKncg5othnm1Rk2PrF?usp=sharing). Upload the 3 files using the [page](https://play.google.com/console/internal-app-sharing) for uploading. Get the 3 links and send them to your device so you can tap on them from there (You could send the links with an email).
3. Uninstall all versions of the Jetpack app that you have on your device.
4. Tap on the first link for version 79999.
5. A Google Play page appears. Tap on the "Install" button. When the download is over, tap on "Open".
6. Login with a **non-automattic** account.
7. From your device, tap the second link for version 80000. You will see a page with an "Update" button. DO NOT tap on it and go back to the app.
- [ ]  Check that nothing happens and no update is shown. 
8. Tap on "Me" button in the bottom navigation bar and tap on "Debug settings".
9. Enable "in app updates" FF. 
10. Kill and restart the app.

- [ ] Check that a full screen appears for the "immediate" update. 
- [ ] Check that the IN_APP_UPDATE_SHOWN event is tracked.
![Screenshot_20240517-125236](https://github.com/wordpress-mobile/WordPress-Android/assets/5176760/4663ecfd-e995-40a9-9e73-5e7990689e5a)

11. Tap on the X button on the top right corner to dismiss it.

- [ ] Check that when close and relaunch the no update screen appears
- [ ] Check that the IN_APP_UPDATE_DISMISSED event is tracked.

12. Wait for **5 minutes or more**. 
14. Launch the app.

- [ ] Check that the Update screen reappears.
15. Tap on Update button.

- [ ] Check that a full screen appears from Google Play lib showing the download progress and then installs the update. When the update is installed the app is relaunched. 
- [ ] Check that the version of the app is 80000 now. 
16. Go and tap the link for version 80001 on your device. You can see a page with an "Update" button. **DO NOT** tap on it and go back to the app. 

- [ ] Check that a bottom sheet appears about the update. 
- [ ] Check that the IN_APP_UPDATE_SHOWN event is tracked
17. Tap on the "X" button of the bottom sheet. 
- [ ]  Check that the IN_APP_UPDATE_DISMISSED event is tracked
- [ ] Check that when you restart the app or put it in the background and then relaunch it, no update bottom sheet appears. (This is because there is an interval of how often a flexible update can appear so we won't annoy our users. This is defined by a remote config and its current value is 5 days.)
18. Clear the data of the app, login again, and enable in-app-updates FF.
19. Kill and restart the app.

- [ ] Check that the update bottom sheet appears.
20. Tap "Update" button. 
- [ ] Check that the IN_APP_UPDATE_ACCEPTED event is tracked. 
- [ ] Check that the update is downloading in the background. 
21. When the update is downloaded and our app is in the foreground: 

- [ ] Check that a snackbar appears that informs the user and asking for a restart. 
22. Tap on the restart action.
- [ ] Check that the IN_APP_UPDATE_COMPLETED_WITH_APP_RESTART event is tracked. 
- [ ] Check that the update is installed and the app is restarted. 

**Attention**: If the app is in the background when the update is downloaded, the update is installed silently without notifying the user.

Please feel free to explore unhappy paths, like turning off the internet connection while an update is being downloaded and then trying to restart the app.



-----

## Regression Notes

1. Potential unintended areas of impact

    - N/A

24. What I did to test those areas of impact (or what existing automated tests I relied on)

    - Manual testing and unit tests.

25. What automated tests I added (or what prevented me from doing so)

    - N/A

-----

## PR Submission Checklist:


-----

## Testing Checklist (strike-out the not-applying and unnecessary ones):
